src/platform/graphics/depth-state.js

```diff
@@ -0,0 +1,149 @@
+ * rendering pipeline. A depth state can be set on a material using {@link Material#depthState},
+ * or in some cases on the graphics device using {@link GraphicsDevice#setDepthState}.
+ *
+ * For the best performance, do not modify depth state after it has been created, but create
+ * multiple depth states and assign them to the material or graphics device as needed.
+ */
+class DepthState {
+    /**
+     * Number, bits of which represent the depth state.
+     *
+     * @private
+     */
+    data = 0;
+
+    /**
+     * Create a new Depth State instance.
+     *
+     * @param {number} func - Controls how the depth of the fragment is compared against the
+     * current depth contained in the depth buffer. See {@link pc.DepthState.func} for details.
+     * Defaults to {@link FUNC_LESSEQUAL}.
+     * @param {boolean} write - If true, shader write a depth value to the depth buffer of the
+     * currently active render target. If false, no depth value is written. Defaults to true.
+     */
+    constructor(func = FUNC_LESSEQUAL, write = true) {
+        this.func = func;
+        this.write = write;
+    }
+
+    /**
+     * If true, shader fragment is only written to the current render target if it passes the depth
+     * test. If false, it is written regardless of what is in the depth buffer. Note that when depth
+     * testing is disabled, writes to the depth buffer are also disabled. Defaults to true.
+     *
+     * @type {boolean}
+     */
+    set test(value) {
+        this.func = value ? FUNC_LESSEQUAL : FUNC_ALWAYS;
+    }
+
+    get test() {
+        return this.func !== FUNC_ALWAYS;
+    }
+
+    /**
+     * If true, shader write a depth value to the depth buffer of the currently active render
+     * target. If false, no depth value is written.
+     *
+     * @type {boolean}
+     */
+    set write(value) {
+        this.data = BitPacking.set(this.data, value ? 1 : 0, writeShift);
+    }
+
+    get write() {
+        return BitPacking.all(this.data, writeShift);
+    }
+
+    /**
+     * Controls how the depth of the fragment is compared against the current depth contained in
+     * the depth buffer. Can be:
+     *
+     * - {@link FUNC_NEVER}: don't draw
+     * - {@link FUNC_LESS}: draw if new depth < depth buffer
+     * - {@link FUNC_EQUAL}: draw if new depth == depth buffer
+     * - {@link FUNC_LESSEQUAL}: draw if new depth <= depth buffer
+     * - {@link FUNC_GREATER}: draw if new depth > depth buffer
+     * - {@link FUNC_NOTEQUAL}: draw if new depth != depth buffer
+     * - {@link FUNC_GREATEREQUAL}: draw if new depth >= depth buffer
+     * - {@link FUNC_ALWAYS}: always draw
+     *
+     * @type {number}
+     */
+    set func(value) {
+        this.data = BitPacking.set(this.data, value, funcShift, funcMask);
+    }
+
+    get func() {
+        return BitPacking.get(this.data, funcShift, funcMask);
+    }
+
+    copy(rhs) {
```

Comment:
Maybe add JSDoc block for `copy`. Not sure if you want it public API or now. I don't see why not though.

Reply:
on it, thanks for the review, much appreciated, I missed so much.